Removes typed endpoint from search and related APIs

colings86 opened the pull request.

colings86 committed:
Removes typed endpoint from search and related APIs

elasticmachine commented:
Pinging @elastic/es-search

colings86 committed:
Removes tpyed endpoints for msearch and mtermvector APIs

colings86 commented:
@elasticmachine run elasticsearch-ci/2

colings86 commented:
@elasticmachine run elasticsearch-ci/2

colings86 commented:
@elasticmachine run elasticsearch-ci/1

colings86 commented:
@elasticmachine run elasticsearch-ci/2

jtibshirani commented on docs/reference/migration/migrate_8_0/reindex.asciidoc:
```diff
@@ -7,4 +7,9 @@ re-encode them when generating the search request for the remote host. This
+[float]
+==== Removal of types
+
+The `/{index}/{type}/_delete_by_query` and `/{index}/{type}/_update_by_query` REST endpoints have been removed in favour of `/{index}/_delete_by_query` and `/{index}/_update_by_query`. Since indexes can now only contain a single unnamed type these typed endpoints are obselete.
```
I wonder if the phrasing 'indexes can now only contain a single unnamed type' could be confusing. We now provide 'typeless' index creation APIs, which to users seem like they don't introduce a type at all. Perhaps we could just say 'Since indexes no longer contain types, these typed endpoints are obsolete.'

Also small typo: obselete -> obsolete.

colings86 replied:
++ good catch, I think your wording is much better. I'll change it in this and the other PR

jtibshirani reviewed (commented):
Just filling in on types-related reviews as @jpountz is out. It is great to see `+72 -1,104 lines`!

I left one comment on the code and had a couple thoughts:
- Since we're removing support for the typed endpoints, using types in certain Java HLRC calls will now result in failures. Are we planning to remove types in HLRC requests/ request conversion code in a future PR, or would it make sense to also cover it here to avoid an 'in-between' state?
- There are a few more search-related areas where types have been deprecated, such as searches on the `_type` field and the use of types in 'lookup' queries like `terms` lookup and percolator. If this PR is already getting fairly large, perhaps we could just make sure to track them on the meta-issue.

colings86 commented:
@elasticmachine run elasticsearch-ci/2

colings86 commented:
Thanks for the review @jtibshirani 

> Since we're removing support for the typed endpoints, using types in certain Java HLRC calls will now result in failures. Are we planning to remove types in HLRC requests/ request conversion code in a future PR, or would it make sense to also cover it here to avoid an 'in-between' state?

I had a look at doing this but unless I'm mistaken `SearchRequest` is one of the remaining classes shared between the HLRC and the server code? So removing types from the SearchRequest very quickly gets into removing types from the search execution in general which is a bigger fix that IMO is suited to solving in its own PR. I added a MUST to https://github.com/elastic/elasticsearch/issues/41059 to remove types from the Search Request and I'm happy to tackle this one straight after this PR is merged to avoid the "in between" state being around for very long which I think is ok since the "in between" state will only exist on master so there is a low risk of releasing in this state. Does this sound ok to you or do you think we really need to tackle these together?

> There are a few more search-related areas where types have been deprecated, such as searches on the _type field and the use of types in 'lookup' queries like terms lookup and percolator. If this PR is already getting fairly large, perhaps we could just make sure to track them on the meta-issue.

I've added a check-box for this in https://github.com/elastic/elasticsearch/issues/41059. I would like to handle this in a separate change to avoid this PR becoming too big

colings86 commented:
@elasticmachine run elasticsearch-ci/1

jtibshirani commented on docs/reference/migration/migrate_8_0/search.asciidoc:
```diff
@@ -0,0 +1,10 @@
+[float]
+==== Removal of types
+
+The `/{index}/{type}/_search`, `/{index}/{type}/_msearch`, `/{index}/{type}/_search/template` and `/{index}/{type}/_msearch/template` REST endpoints have been removed in favour of `/{index}/_search`, `/{index}/_msearch`, `/{index}/_search/template` and `/{index}/_msearch/template`, since indexes no longer contain types, these typed endpoints are obsolete..
```
Accidental double period here (and below).

jtibshirani reviewed (approved):
> I'm happy to tackle this one straight after this PR is merged to avoid the "in between" state being around for very long

Tackling it in the next PR sounds good to me, thanks! I agree there's not a perfect split into PRs because we share the server and HLRC classes in some places.

colings86 commented:
@elasticmachine run elasticsearch-ci/docbldesx

colings86 commented:
@elasticmachine run elasticsearch-ci/1

colings86 commented:
@elasticmachine run elasticsearch-ci/packaging-sample
